meson.build

```diff
@@ -287,23 +287,41 @@ endif
 # Compression support
 #-------------------------------------------------------------------------------
 
-zstd = dependency('libzstd', version : '>= 1.4.4', required : get_option('zstd'))
+zstd_dlopen = get_option('dlopen').contains('zstd')
+if get_option('zstd').disabled() and zstd_dlopen
+  error('Incompatible options: dlopen\'ed zstd needs zstd at build time')
+endif
+zstd = zstd_dlopen ? dependency('libzstd', version : '>= 1.4.4', required : true) \
+                   : dependency('libzstd', version : '>= 1.4.4', required : get_option('zstd'))
 if zstd.found()
   cdata.set('ENABLE_ZSTD', true)
+  cdata.set10('ENABLE_DLOPEN_ZSTD', zstd_dlopen)
```

Comment:
`ENABLE_DLOPEN_ZSTD` is not defined in the `not zstd.found()` case. Which should result in build failure due to the `Wundef` we (rightfully) have.

Reply:
Fixed and also changed the rest of the build system to always define the ENABLE_* macros